### PR TITLE
Pod/Container can steal cpu/memory resources if cluster admin did not set minimum (not maximum) limitrange

### PR DESCRIPTION
**Pull request by mburke5678**
https://bugzilla.redhat.com/show_bug.cgi?id=1644512

### CODE REVIEW

**Comment by mburke5678** [comment]
@sjenning @nak3  PTAL
This issue is related only to the CPU and memory settings in LimitRange objects only and not other CPU/Memory settings?

File: admin_guide/limits.adoc [hunk 1]
```diff
@@ -172,7 +189,8 @@ Per container, the following must hold true if specified:
 (optional)
 
 If the configuration defines a `min` CPU, then the request value must be greater
-than the CPU value. A limit value does not need to be specified.
+than the CPU value. Not setting a `min` value or setting `0` is unlimited 
+allowing the pod to consume more than than  the `max` value.
```
Comment (nak3) [inline admin_guide/limits.adoc]: `than than` must be a typo.

File: admin_guide/limits.adoc [hunk 2]
```diff
@@ -224,17 +242,18 @@ Across all containers in a pod, the following must hold true:
 
 |`*Min*`
 |`Min[resource]` less than or equal to `container.resources.requests[resource]`
-(required) less than or equal to `container.resources.limits[resource]`
-(optional)
+(required) less than or equal to `container.resources.limits[resource]`.
+Not setting a `min` value or setting `0` is unlimited allowing the 
+pod to consume more than than  the `max` value.
```
Comment (nak3) [inline admin_guide/limits.adoc]: Same with above `than than`.

**Comment by nak3** [comment]
Thank you @mburke5678 

>  For CPU and Memory limits, if you specify a `max` value, but do not specify a
> `min` limit, the resource can consume CPU/memory resources greater than `max` value`.

This issue needs following two conditions:

- a. Admin does not specify `min` limit or does specify `0` in `Limitrange`.
- b. Users specify `0` in `resources.limits.{cpu,memory}` in `Pod`.

The fix does not mention about b, but do you have any idea how we can explain about b? (I am wondering that current fix is also OK, but if you have any good idea, please let me know.)

**Comment by nak3** [comment]
> Which OpenShift versions does this apply to? 3.7+? 4.1?

All OpenShift and Kuberntes versions have this issue.

**Comment by nak3** [comment]
lgtm, but defer to @sjenning

**Comment by mburke5678** [comment]
Is it the case that setting `min: cpu:` and `min:memory` to `0` in a pod spec mean unlimited and allow the pod to consume more than `max` in all cases, or only when a LimitRange object exist? 
Either way, it might be worth mention in the docs. 
I did a quick search on `cpu:` and found 70 references in 22 files, not all of which are about setting the `cpu` and `memory` values. 
Just want a second opinion on whether we should add such a note in all situations where the user is setting these values for a pod; or mention once to the admin and let the admin tell users?

**Comment by mburke5678** [comment]
@sjenning 
Is the LimitRange object in 4.x? It doesn't seem we brought that topic forward from 3.11 docs.

**Comment by mburke5678** [comment]
@sjenning 
Is it the case that setting min: cpu: and min:memory to 0 in a pod spec mean unlimited and allow the pod to consume more than max in all cases, or only when a LimitRange object exist?
Either way, it might be worth mention in the docs.
I did a quick search on cpu: and found 70 references in 22 files, not all of which are about setting the cpu and memory values.
Just want a second opinion on whether we should add such a note in all situations where the user is setting these values for a pod; or mention once to the admin and let the admin tell users?

**Comment by mburke5678** [comment]
@sjenning PTAL

**Comment by sjenning** [comment]
> Is the LimitRange object in 4.x? It doesn't seem we brought that topic forward from 3.11 docs.

yes

> Is it the case that setting min: cpu: and min:memory to 0 in a pod spec mean unlimited and allow the pod to consume more than max in all cases, or only when a LimitRange object exist?

setting them both to 0 means the pod is best effort and can consume as much cpu and memory as the system makes available best effort pods (cpu scraps and memory not used by guaranteed pods).  The limit range can 1) not exist 2) exist but have no min 3) exist and have the min set to 0 and this is possible.

LGTM

**Pull request merged**